fedimint-atomic-broadcast/src/lib.rs

```diff
@@ -0,0 +1,246 @@
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Recipient {
+    Everyone,
+    Peer(PeerId),
+}
+
+/// This enum specifies wether an [OrderedItem] has been accepted or discarded
+/// by Fedimint Consensus.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Decision {
+    Accept,
+    Discard,
+}
+
+/// If two correct nodes obtain two ordered items from the broadcast they
+/// are guaranteed to be in the same order. However, an ordered items is
+/// only guaranteed to be seen by all correct nodes if a correct node decides to
+/// accept it.
+#[derive(Debug)]
+pub struct OrderedItem {
+    pub item: Vec<u8>,
+    pub block_index: u64,
+    pub peer_id: PeerId,
+    pub decision_sender: oneshot::Sender<Decision>,
+}
+
+/// An accepted item is created and recorded in the corresponding block if and
+/// only if  Fedimint Consensus decides to accept an [OrderedItem].
+#[derive(Clone, Debug, PartialEq, Eq, Encodable, Decodable)]
+pub struct AcceptedItem {
+    pub item: Vec<u8>,
+    pub peer_id: PeerId,
+}
+
+/// All items ordered in a session that have been accepted by Fedimint Consensus
+/// are recorded in the corresponding block. A running Federation produces a
+/// [Block] roughly every five minutes.  Therefore, just like in Bitcoin, a
+/// [Block] might be empty if no items are ordered in that time or all ordered
+/// items are discarded by Fedimint Consensus.
+#[derive(Clone, Debug, PartialEq, Eq, Encodable, Decodable)]
+pub struct Block {
+    pub index: u64,
+    pub items: Vec<AcceptedItem>,
+}
+
+impl Block {
+    /// A blocks header consists of 40 bytes formed by its index in big endian
+    /// bytes concatenated with the merkle root build from the consensus
+    /// hashes of its [AcceptedItem]s or 32 zero bytes if the block is
+    /// empty. The use of a merkle tree allows for efficient inclusion
+    /// proofs of accepted consensus items for clients.
+    pub fn header(&self) -> [u8; 40] {
+        let mut header = [0; 40];
+
+        header[..8].copy_from_slice(&self.index.to_be_bytes());
+
+        let leaf_hashes = self.items.iter().map(|item| {
+            let mut engine = sha256::HashEngine::default();
+            item.consensus_encode(&mut engine)
+                .expect("Writing to HashEngine cannot fail");
+            sha256::Hash::from_engine(engine)
+        });
+
+        if let Some(root) = merkle_tree::calculate_root(leaf_hashes) {
+            header[8..].copy_from_slice(&root.to_byte_array());
+        }
+
+        header
+    }
+}
+
+/// A signed block combines a block with the naive threshold secp schnorr
+/// signature for its header created by the federation. The signed blocks allow
+/// clients and recovering guardians to verify the federations consensus
+/// history. After a signed block has been created it is stored in the database.
+#[derive(Clone, Debug, Encodable, Decodable)]
+pub struct SignedBlock {
+    pub block: Block,
+    pub signature: std::collections::BTreeMap<PeerId, [u8; 64]>,
```

Comment:
signatureS maybe? Maybe `peer_signatures` would be most informative.

Reply:
Done